### ISSUE TITLE
SonarQube C++ Community Plugin Review

### ISSUE THREAD

**Issue by guwirth** [SonarOpenCommunity/sonar-cxx#552]
**Conclusion**
---
**Pros**
* Support of a lot of external tools
* Very easy to use
* Duplicated code analysis
* Very good code coverage analysis integration

**Cons**
* Too few integrated inspections
* Limited parsing of C++
* Not fully compatible with C++11/C++14
* False positives
* Not enough love for clang (compiler warnings, clang-tidy, tooling, static-analyzer, ...)

You can read the full review from Baptiste Wicht here:
http://baptiste-wicht.com/posts/2015/05/sonar-cpp-community-plugin-review.html

**Comment by jmecosta** [comment]
the question here how much effort is required to support all c++ standards
by using sslr? we really should consider if using clang as squid as
replacement is or not a better alternative. ive created a fsharp-plugin
using a similar approach as csharp and the effort to set it up was not that
great. see https://github.com/jmecosta/sonar-fsharp-plugin it uses the f#
compiler services so no need to worry about creating a language parser

**Comment by guwirth** [comment]
Hello @jmecosta. In general and mid-term I agree see #272. But short-term with the goal of a 1.0 I think we should not change our strategie. We are very close to have full C++11 and C++14 support and also the most important compiler extensions from gcc and MSVC are supported.

**Comment by jmecosta** [comment]
Yep I also agree with you for 1.0

**Comment by jens-auer** [comment]
What would be the way to integrate clang into the plugin? Would you extend the already existing clang static analyzer and parse the output, or would you use clang as a library in the sonar plugin?

**Comment by guwirth** [comment]
> Would you extend the already existing clang static analyzer and parse the output,

Short term this is the preferred way, extending existing ```sonar.cxx.compiler```.

> or would you use clang as a library in the sonar plugin?

See the ongoing discussion on #272

**Comment by guwirth** [comment]
@wichtounet There is improved C++ compliance now: https://github.com/wenns/sonar-cxx/releases/tag/cxx-0.9.4-RC3

**Comment by wichtounet** [comment]
I just updated my plugin to the 0.9.4-RC3 version. I have only tested it on my biggest project now, but I haven't seen any difference, I still have the same issues with the type aliases. I'll see with the analysis of the other project if I see a difference. 

Do you want me to test something specific ?

**Comment by guwirth** [comment]
@wichtounet Thought after adding a lot of C++11 and C++14 features you should get better results.

> I still have the same issues with the type aliases

Could you provide concrete examples that we can fix it please.

**Comment by HMadhu** [comment]
Hello guys,
Sonar is very new to me.. I am using sonar 5.1.1 and sonar runner 4.2 and community plugin v0.9.3 for analysing c++/c projects.. I followed the documentation and I generated .xml using cppcheck v1.69.. I copied them in my project source directory and ran sonar runner on the projects root directory.. In the dashboard I could see 0 issue and technical debt also 0 and coverage also 0... I tried almost all possible way but nothing is working.. please help me to resolve this issue...

**Comment by jmecosta** [comment]
Plead open a new issue.

**Comment by guwirth** [comment]
@wichtounet I like to fix your type alias problem:
http://en.cppreference.com/w/cpp/language/type_alias
Currently we are testing:
https://github.com/wenns/sonar-cxx/blob/master/cxx-squid/src/test/resources/parser/own/C%2B%2B11/using-alias.cc

What is different in your case?

**Comment by wichtounet** [comment]
Using a 0.9.4-SNAPSHOT (some months old), here are some examples of false positives: 

```c++
template <typename RBM, typename Trainer>
void update_normal(RBM& rbm, Trainer& t) {
    using rbm_t = RBM;
    // Issue: At most one statement is allowed per line, but 2 statements were found on this line. 
```

```c++
template <typename RBM, typename Trainer>
void update_convolutional(RBM& rbm, Trainer& t) {
    using rbm_t  = RBM;
    //At most one statement is allowed per line, but 2 statements were found on this line.

    using weight = typename rbm_t::weight;
    //At most one statement is allowed per line, but 2 statements were found on this line. 
```

```c++
template <typename Iterator, typename LIterator>
typename dbn_t::weight train(DBN& dbn, Iterator first, Iterator last, LIterator lfirst, LIterator llast, std::size_t max_epochs) const {
        constexpr const auto batch_size     = std::decay_t<DBN>::batch_size;
        constexpr const auto big_batch_size = std::decay_t<DBN>::big_batch_size;

        using samples_t = std::vector<typename std::iterator_traits<Iterator>::value_type>;
        //At most one statement is allowed per line, but 2 statements were found on this line.

        using labels_t  = std::vector<typename std::iterator_traits<LIterator>::value_type>;
        //At most one statement is allowed per line, but 2 statements were found on this line. 
```

You can see examples online: https://baptiste-wicht.ch/sonar/drilldown/issues/1179?rule=cxx%3ATooManyStatementsPerLine&rule_sev=MAJOR

**Comment by guwirth** [comment]
@wichtounet I tried your examples with the sslr-toolkit 0.9.4-RC4. There I get no errors. Could you please try on your own machine with https://github.com/wenns/sonar-cxx/releases/download/cxx-0.9.4-RC4/sslr-cxx-toolkit-0.9.4-SNAPSHOT.jar. (In toolkit text must be UTF-8. Copy example to tool like notepad and back into toolkit):

```Java
template <typename RBM, typename Trainer>
void update_normal(RBM& rbm, Trainer& t) {
    using rbm_t = RBM;
    //At most one statement is allowed per line, but 2 statements were found on this line.
} 

template <typename RBM, typename Trainer>
void update_convolutional(RBM& rbm, Trainer& t) {
    using rbm_t  = RBM;
    //At most one statement is allowed per line, but 2 statements were found on this line.

    using weight = typename rbm_t::weight;
    //At most one statement is allowed per line, but 2 statements were found on this line. 
}

template <typename Iterator, typename LIterator>
typename dbn_t::weight train(DBN& dbn, Iterator first, Iterator last, LIterator lfirst, LIterator llast, std::size_t max_epochs) const {
        constexpr const auto batch_size     = std::decay_t<DBN>::batch_size;
        constexpr const auto big_batch_size = std::decay_t<DBN>::big_batch_size;

        using samples_t = std::vector<typename std::iterator_traits<Iterator>::value_type>;
        //At most one statement is allowed per line, but 2 statements were found on this line.

        using labels_t  = std::vector<typename std::iterator_traits<LIterator>::value_type>;
        //At most one statement is allowed per line, but 2 statements were found on this line. 
}
```

**Comment by wichtounet** [comment]
@guwirth I have put one the source file into the tool, but I don't know what I'm supposed to do with the tool. I clicked on Parse and then the AST appeared, but it does not have the sources and all the context. I don't see any error.

**Comment by wichtounet** [comment]
Just to make sure, I downloaded the RC4 and the problem is still the same.

**Comment by guwirth** [comment]
@wichtounet The  sslr-toolkit is using the C++ parser to create an Abstract Syntax Tree (AST). On the left side (Source Code) of the tool you can insert/type code and verify it by pressing the 'Parse Source Code' button. In case you get an 'Abstract Syntax Tree' the parser is supporting the code. Otherwise you see an exception text in the right window.

Your sample code (at least the code snippet above) is supported without an error. So this snippets can't be the root cause for the problem. Just to be on the save side: Do you also get an AST with the code above?

**Comment by guwirth** [comment]
@wichtounet now I think I've got your point. You get a rule violation by the check https://github.com/wenns/sonar-cxx/blob/master/cxx-checks/src/main/java/org/sonar/cxx/checks/TooManyStatementsPerLineCheck.java in case of a type alias (https://baptiste-wicht.ch/sonar/drilldown/issues/1179?rule=cxx%3ATooManyStatementsPerLine&rule_sev=MAJOR).

**Comment by guwirth** [comment]
Close this issue tracked in #656

**Comment by wichtounet** [comment]
I tested the new release and the false positives did go away :)

Now I only have issues coming from the test framework :)

Thanks guys!
